extendr-api/src/wrapper/array.rs

```diff
@@ -0,0 +1,339 @@
+    Robj: AsTypedSlice<'a, T>,
+{
+    type Error = Error;
+
+    fn try_from(robj: &'a Robj) -> Result<Self> {
+        Self::try_from(robj.clone())
+    }
+}
+
+impl<T, D> Deref for RArray<T, D> {
+    type Target = Robj;
+
+    fn deref(&self) -> &Self::Target {
+        &self.robj
+    }
+}
+
+const BASE: usize = 0;
+
+trait Offset<D> {
+    /// Get the offset into the array for a given index.
+    fn offset(&self, idx: D) -> usize;
+}
+
+impl<T> Offset<[usize; 1]> for RArray<T, [usize; 1]> {
+    /// Get the offset into the array for a given index.
+    fn offset(&self, index: [usize; 1]) -> usize {
+        if index[0] - BASE > self.dim[0] {
+            panic!("array index: row overflow");
+        }
+        index[0] - BASE
+    }
+}
+
+impl<T> Offset<[usize; 2]> for RMatrix<T> {
+    /// Get the offset into the array for a given index.
+    fn offset(&self, index: [usize; 2]) -> usize {
+        if index[0] - BASE > self.dim[0] {
+            panic!("matrix index: row overflow");
+        }
+        if index[1] - BASE > self.dim[1] {
+            panic!("matrix index: column overflow");
+        }
+        (index[0] - BASE) + self.dim[0] * (index[1] - BASE)
+    }
+}
+
+impl<'a, T: 'a> Index<[usize; 2]> for RMatrix<T>
+where
+    Robj: AsTypedSlice<'a, T>,
+{
+    type Output = T;
+
+    /// Zero-based indexing in row, column order.
+    ///
+    /// Panics if out of bounds.
+    /// ```
+    /// use extendr_api::prelude::*;
+    /// test! {
+    ///     let matrix = RArray::from_slice([1.,2.,3.,4.,5.,6.], [3, 2]).unwrap();
+    ///     assert_eq!(matrix[[0, 0]], 1.);
+    ///     assert_eq!(matrix[[1, 0]], 2.);
+    ///     assert_eq!(matrix[[2, 1]], 6.);
+    /// }
+    /// ```
+    fn index(&self, index: [usize; 2]) -> &Self::Output {
+        let ptr = self.robj.as_typed_slice().unwrap().as_ptr();
+        unsafe { ptr.add(self.offset(index)).as_ref().unwrap() }
+    }
+}
+
+impl<'a, T: 'a> IndexMut<[usize; 2]> for RMatrix<T>
+where
+    Robj: AsTypedSlice<'a, T>,
+{
+    /// Zero-based mutable indexing in row, column order.
+    ///
+    /// Panics if out of bounds.
+    /// ```
+    /// use extendr_api::prelude::*;
+    /// test! {
+    ///     let mut matrix = RArray::from_slice([0.,0.,0.,0.,0.,0.], [3, 2]).unwrap();
+    ///     matrix[[0, 0]] = 1.;
+    ///     matrix[[1, 0]] = 2.;
+    ///     matrix[[2, 0]] = 3.;
+    ///     matrix[[0, 1]] = 4.;
+    ///     assert_eq!(matrix.as_real_slice().unwrap(), &[1., 2., 3., 4., 0., 0.]);
+    /// }
+    /// ```
+    fn index_mut(&mut self, index: [usize; 2]) -> &mut Self::Output {
+        let ptr = self.robj.as_typed_slice_mut().unwrap().as_mut_ptr();
+        unsafe { ptr.add(self.offset(index)).as_mut().unwrap() }
+    }
+}
+
+mod tests {
+    use super::*;
+
+    #[test]
+    fn from_slice() {
+        test! {
+            // Create 2D RArray.
+            let slice = [1,2,3,4,5,6];
+            let dim = [2,3];
+            let array = RArray::from_slice(slice, dim).unwrap();

```

Comment:
Are there other construction methods besides `from_slice()`?

Reply:
There's also `from_slice_unchecked`. Since creating a new `RArray` creates a copy of the inner data, these were the only 2 methods I thought were needed. Note that you can use `from_slice()` directly on a `Vec` or on a rust array in this case.
I made `RArray::new()` available only to be used internally, since it doesn't add the `dim` attribute for the Robj.
Let me know if you are thinking of another type of constructor.

Reply:
Is there a way to create one without copying data? That would be quite nice. Additionally, it would be handy to be able to determine the col/row major order or at least provide a closure similar to the matrix method.

Reply:
I don't think it is possible to create an `Robj` without copying data.

Is it ok for you if I add an example on how to build row major matrices? Or is the `RMatrix` constructor using a closure really desirable? I removed it because think the data manipulation should be left for the user to figure out and also because it was a somewhat complicated function to understand how to use... 

Maybe @CGMossa and @Ilia-Kosenkov  can give their opinion on this.
```
    #[test]
    fn test_collect_rarray_matrix_row_major() {
        test! {
            // Test building an RMatrix in row major order.
            let v: Vec<i32> = (1i32..=12).collect();
            let nrow = 4_usize;
            let ncol = 3_usize;
            let rmat = (0..ncol)
                .flat_map(|i| {
                    v.iter()
                        .skip(i)
                        .step_by(ncol)
                        .cloned()
                })
                .collect_rarray([nrow, ncol]);
            assert!(rmat.is_ok());
            assert_eq!(Robj::from(rmat), R!("matrix(1:12, nrow=4, byrow = TRUE)").unwrap());
        }
    }
```

Reply:
I would say that the closure is very important. I can find 8 examples where I'm using it in existing code. 
https://github.com/search?q=owner%3AJosiahParry%20new_matrix&type=code

It's very helpful in that I often iterate over rows and store a `Vec<T>` where `T` is some structure that is on a per-row basis. Presently I am writing code that creates a matrix from `Vec<[f64;2]>` with the closure `|r, c| x[r][c]`.

Reply:
The closure is definitely useful, but I find it very confusing. That's why I removed it in the first place. The changes I made are very opinionated as I think these transformations made by the closure should be done outside extendr. When creating the array, I think the user should already have the Vec (or slice) already ready to be converted to an RArray.

Reply:
It requires a whole lot of mental gymnastics to understand how to create a flattened vector in the appropriate shape for R's column-major matrices. Not many people will be able to figure that out--myself included. QOL >>> imo